In [1]:
# setup
from IPython.core.display import display,HTML
display(HTML('<style>.prompt{width: 0px; min-width: 0px; visibility: collapse}</style>'))
display(HTML(open('rise.css').read()))

# imports
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set(style="whitegrid", font_scale=1.5, rc={'figure.figsize':(12, 6)})


# CMPS 2200
# Introduction to Algorithms

## Functional Programming


## Lambda Calculus

- pure language developed by Alonzo Church in the 1930s
- inherently parallel

Consists of expressions $e$ in one of three forms:

1. a **variable**, e.g., $x$
2. a **lambda abstraction**, e.g., $(\lambda \: x \: . \: e)$, where $e$ is a function body.
3. an **application**, written $(e_1, e_2)$ for expressions $e_1$, $e_2$.

### Beta reduction

A generic way of applying a sequence of anonymous functions.

$$
(\lambda \: x \: . \: e_1) e_2 \mapsto e_1 [x / e_2]
$$

$e_1 [x / e_2]$: for every free occurence of $x$ in $e_1$, substitute it with $e_2$.

Can read as "substitute $e_2$ for $x$ in $e_1$."

E.g.


In [11]:
# lambda functions exist in Python.
# these are anonymous functions (no names)
# Here, e_2 is a variable.
(lambda x: x*x)(10)

100


$
(\lambda \: x \: . \: x * x) \:  10 \mapsto 10 * 10 \mapsto 100
$

We can also chain functions together. E.g., $e_2$ can be another function.

In [12]:
(lambda x: x*x)((lambda x: x+2)(10))

144

beta reduction:

$ (\lambda \: x \: . \: x * x)((\lambda \: x \: . \:   x+2) \: 10 ) \mapsto$  
$(\lambda \: x \: . x * x) \: (10+2) \mapsto$  
$(\lambda \: x \: . x * x) \: 12 \mapsto$  
$( 12 * 12 ) \mapsto$  
$ 144 $

<br>

**Could we have done this in any other order?**

$( (\lambda \: x \: . \: x * x) (\lambda \: x \: . \:   x+2)) \: 10  \mapsto$  
$(\lambda \: x \: . (x+2) * (x+2)) \: 10 \mapsto$  
$(10+2) * (10+2) \mapsto$  
$ 144 $


## Beta reduction order

$$
(\lambda \: x \: . \: e_1) e_2 \mapsto e_1 [x / e_2]
$$

**call-by-value:** (first example): $e_2$ is evaluated to a value first, then reduction is applied
- square(10+2)
- square(12)
- 12*12
- 144
- Languages that use call-by-value are called **strict**: argument is always evaluated before applying function



**call-by-need:** (second example): $e_2$ is first copied into $e_1$, then $e_1$ is evaluated.
- square(10+2)
- (10+2)*(10+2)
- 144


We will be using call-by-value, which is more amenable to parallelism.

## Computation via beta reductions

**Computation** in lambda calculus means to apply beta reductions until there is nothing left to reduce.

When there is nothing left to reduce, the expression is in **normal form**.

<br>

How can we make an infinite loop in lambda computation?

$ ( (\lambda \: x \: . \: (x \: x)) (\lambda \: x \: . \: (x \: x))) \mapsto$

$ ( (\lambda \: x \: . \: (x \: x)) (\lambda \: x \: . \: (x \: x)))$

Doing computation via lambda calculus seems limiting. 

Can it compute everything we need?

Surprising result:

The Lambda calculus is equivalent to Turing Machines.

That is...  

Anything that can be computed by a Turing Machine can also be computed by the Lambda calculus, and visa versa.


More details: 
- CMPS/MATH 3250
- [**Church-Turing Thesis**](https://en.wikipedia.org/wiki/Church%E2%80%93Turing_thesis)
